Add New Layerpicker

arottersman opened the pull request:
## Overview
This PR aims to take us 80% of the way to having a new, wired up layer picker. 

- The first commit is Jeff's prototype
- The second commit reorganizes how we store the layer config. Originally it was a long list of all the layers, and used boolean flags to determine what category they belonged to. Now they have been broken up into a dictionary. Each key holds a list with all the layers of that type (eg streams, basemap, etc)
- The third commit adds the new layer picker control as its own set of backbone views (instead of a leaflet control). The building of layers from the config is now handled by a layers backbone model.

#### Todo
- [x]  ~Opacity Control~ PR #1667 
- [x] ~Observations Deferred Tab (don't click the final tab right now...)~ PR #1673 
- [x] ~Legends~ issue #1674
- [x] Testing ITSI Interactive Screenshot (changed some code related to fixing the google layer bug #1640)
- [x] Compare view bug (when you come back from the compare view, you get stuck with a single tile until you switch layers/zooms. See screenshot below)
<img width="960" alt="screen shot 2017-01-17 at 6 24 03 pm" src="https://cloud.githubusercontent.com/assets/7633670/22070556/75f6ff0e-dd6a-11e6-8b44-e529a748324f.png">

## Demo
![layerpickeruno](https://cloud.githubusercontent.com/assets/7633670/22070478/300c3144-dd6a-11e6-904f-559e76cc95b7.gif)

## Testing Instructions
- Pull this branch and bundle up
- Add and remove layers. 
- Confirm certain layers (like PA municipalities) are disabled when you are too zoomed out, or the viewport is outside the US. 
- Test out the Select By Boundary

Connects #1652

arottersman commented:
Fixup addresses https://github.com/WikiWatershed/model-my-watershed/issues/1652#issuecomment-273486021, and puts boundary and coverage grid layers in their own tabs.

I've left the implementation, however, whereby you could have multiple layer groups in one tab. Not sure if this is desirable?

Here's what the layer picker looks like with 5 tabs, but the same overall width:
![screen shot 2017-01-18 at 11 36 38 am](https://cloud.githubusercontent.com/assets/7633670/22073104/66dbdf78-dd72-11e6-8fd9-16d7203b3df7.png)

arottersman commented:
Realized the compare view bug was a pretty small change, so added a commit to fix it on this branch. Will create new branches/PRs for the subsequent todos.

mmcfarland commented on src/mmw/js/src/core/layerPicker.js:
```diff
@@ -0,0 +1,274 @@
+
+    updateDisabled: function(layer, shouldDisable) {
+        this.collection.findWhere({ display: layer.display })
+            .set('disabled', shouldDisable);
```
I'm not seeing this `disabled` property translate to the rendered template for `minZoom` invalidations, I think this feature got dropped.  I can show you quickly what the expected behavior is.

arottersman commented:
This should be ready for another look. 

- Fixed the disable on zoom, and added `disabled` styling to make it easier to test:
![screen shot 2017-01-23 at 11 10 46 am](https://cloud.githubusercontent.com/assets/7633670/22211877/9c7d5272-e15c-11e6-9b28-48832e3d0d59.png)

- For the weird rendering on your machine, I think we should spin that off to a "finalize layer picker styling" sort of issue. It renders similarly on IE11 (see #1667), and I'm sure there are other tweaks @jfrankl will want to make now that he can play with it all wired up.

mmcfarland commented:
@azavea-bot rebuild

arottersman commented on src/mmw/apps/home/views.py:
```diff
@@ -103,12 +104,17 @@ def get_model_packages():
     layer_config_with_urls = deepcopy(settings.LAYER_GROUPS)
-    [[layer.update({'url': get_layer_url(layer)}) for layer in layers
+    layers = list(itertools.chain.from_iterable(
+        layer_config_with_urls.itervalues()))
```
The values of the layer config are lists, so`.values()` returns a list of lists. I'm using the cast and iterator to flatten it.

I can change this to `create_layer_config_with_urls(layer_type)` and call it for each layer type (`base_maps`, `streams`, etc). That way we avoid having to deal with the list of lists at all.

kdeloach commented on src/mmw/js/src/app.js:
```diff
@@ -67,6 +73,20 @@ var App = new Marionette.Application({
+    },
+
+    showLayerPicker: function() {
+        if (_.contains(settings.get('map_controls'), 'LayerSelector')) {
```
Adding more settings helpers could improve how this reads (ex. `settings.isLayerSelectorEnabled`).

arottersman replied:
Do you mean remove `LayerSelector` from the `map_controls` setting and have it standalone?

kdeloach replied:
I mean something like this:

(in settings.js)

```
function isLayerSelectorEnabled() {
    return _.contains(settings.get('map_controls'), 'LayerSelector');
}
```

Just to make this line of code somewhat self-documenting.

arottersman replied:
Gotcha, good idea!

kdeloach commented on src/mmw/js/src/app.js:
```diff
@@ -131,9 +151,8 @@ function initializeShutterbug() {
-            var mapView = App.getMapView(),
-                activeBaseLayer = mapView.baseLayers[mapView.getActiveBaseLayerName()],
-                googleLayerVisible = !!activeBaseLayer._google;
+            var activeBaseLayer = App.getLayersModel().baseLayers.findWhere({ active: true }),
```
To encapsulate implementation details (and since this same logic is repeated elsewhere) you might consider adding a helper method to `LayersModel` to retrieve the active base layer.

kdeloach commented on src/mmw/js/src/core/models.js:
```diff
@@ -83,6 +85,70 @@ var MapModel = Backbone.Model.extend({
+            // Check to see if the google api service has been loaded
+            // before creating a google layer
+            if (layer.googleType){
+                if(self._googleMaps) {
```
Missing space.

kdeloach commented on src/mmw/js/src/core/views.js:
```diff
@@ -360,7 +308,7 @@ var MapView = Marionette.ItemView.extend({
         // The max available zoom level changes based on the active base layer
         this._leafletMap.on('baselayerchange', this.updateCurrentZoomLevel);
-        this._leafletMap.on('baselayerchange', this.updateDrbLayerZoomLevel);
+        this._leafletMap.on('baselayerchange', _.bind(this.updateDrbLayerZoomLevel, this));
```
I think you can pass `context` using the third (optional) argument.

kdeloach commented on src/mmw/js/src/draw/views.js:
```diff
@@ -122,7 +122,7 @@ function validatePointWithinDataSourceBounds(latlng, dataSource) {
             return d.promise();
     }
-
+    
```
Whitespace.

arottersman commented:
- Fixed tests
- Squashed the old fixups and added a new one to address Kevin's comments.

arottersman commented:
Let me know if either of you want to review this again before I merge

kdeloach commented:
Looks good. Don't forget to squash c0f831a0539b3a362018a8a2e24b9d3a2d39a855.

mmcfarland commented:
Yep, looks good.  Nice additional changes.

arottersman commented:
@azavea-bot rebuild
